utils/src/main/java/software/amazon/awssdk/utils/async/SimplePublisher.java

```diff
@@ -401,16 +373,39 @@ public void cancel() {
             log.trace(() -> "Received cancel()");
 
             // Create exception here instead of in supplier to preserve a more-useful stack trace.
-            IllegalStateException failure = new IllegalStateException("A downstream publisher has cancelled the subscription.");
-            rejectException.compareAndSet(null, () -> failure);
-            eventQueue.add(new CancelQueueEntry<>());
-            entryTypesToFail.addAll(asList(ON_NEXT, ON_COMPLETE, ON_ERROR));
+            highPriorityQueue.add(new CancelQueueEntry<>());
             processEventQueue();
         }
     }
 
     /**
-     * An entry in the {@link #eventQueue}.
+     * A lazily-initialized failure message for future events sent to this publisher after a terminal event has
+     * occurred.
+     */
+    private static final class FailureMessage {
+        private Supplier<Throwable> failureMessageSupplier;
+        private Throwable failureMessage;
```

Comment:
Can they be set by different threads? My guess is no, but wanted to double check

Reply:
This is complicated. 

Yes, different threads read and write this value, but that's okay.

It's guaranteed that every write/read happens while holding the processingQueue flag, except one read we make in processEventQueue when it calls shouldProcessQueueEntry. It's okay that a thread might read a stale value there, because it will only happen when there's another thread currently holding the processingQueue flag, and that thread will re-check the value when it releases the flag.